crypto/fipsmodule/kdf/kbkdf.c

```diff
@@ -0,0 +1,95 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/kdf.h>
+#include "internal.h"
+
+int KBKDF_ctr_hmac(uint8_t *out_key, size_t out_len, const EVP_MD *digest,
+                   const uint8_t *secret, size_t secret_len,
+                   const uint8_t *info, size_t info_len) {
+  int ret = 0;
+
+  HMAC_CTX *hmac_ctx = NULL;
+
+  if (!out_key || out_len == 0 || !secret || secret_len == 0) {
+    goto err;
+  }
+
+  hmac_ctx = HMAC_CTX_new();
+  if (!hmac_ctx) {
+    goto err;
+  }
+
+  if (!HMAC_Init_ex(hmac_ctx, secret, secret_len, digest, NULL)) {
+    goto err;
+  }
+
+  // Determine the length of the output in bytes of a single invocation of the
+  // HMAC function.
+  size_t h_output_bytes = HMAC_size(hmac_ctx);
+  if (!h_output_bytes || h_output_bytes <= 0) {
+    goto err;
+  }
+
+  // NIST.SP.800-108r1-upd1: Step 1:
+  // Determine how many output chunks are required to produce the requested
+  // output length |out_len|. This determines how many times the variant compute
+  // function will be called to output key material.
+  uint64_t n = (out_len + h_output_bytes - 1) / h_output_bytes;
```

Comment:
just curious, `out_len` and `h_output_bytes` are `size_t`. If `size_t` is 32 bits wide, when will cast to `uint64_t` happen?

Reply:
Fixed this to compute `n` as a size_t. Then the `uint32_t i` in the for loop will be lifted through implicit conversion when comparing to `n` to size_t for the loop condition expression.

Reply:
I think that won't work for a 32-bit build where `size_t` is 32 bits wide. For example, `out_len` might be equal to `UINT32_MAX`, and `n` won't be computed correctly then (and also the condition `n > UINT32_MAX` wouldn't do anything). I think `n` must be `uint64_t`, I was just wondering if we have to explicitly cast to `uint64_t` any of the variables in the expression:
```
(out_len + h_output_bytes - 1) / h_output_bytes;
```

also, maybe worth thinking about adding a test case for this.